### PR TITLE
[occm] Make volume mounts configurable in Helm Chart

### PR DESCRIPTION
**Pull request by jacksgt**
**What this PR does / why we need it**:
Depending on which controllers of the OCCM are being used, not all
volume mounts specified in the Helm chart are necessary.
Furthermore, some systems use different paths for these components [1] or
they don't have them at all [2], which will result in failure to attach
the volume to the pod and subsequently run the pod.

This patch makes the volume mounts configurable for the Helm chart
user. By default, the same volumes as before (flexvolume-dir,
ca-certs, k8s-certs) are mounted into the pods.
This makes this change transparent to current users of the Helm
chart (no breaking change).

This patch also updates the deprecated --address parameter to use
--bind-address instead.

**Which issue this PR fixes(if applicable)**: N/A

**Special notes for reviewers**:
I performed manual testing with these changes, i.e. using `helm template occm charts/openstack-cloud-controller-manager/` to verify that 1) the output is as expected and 2) the generated YAML manifests are the same as before (no breaking changes).

**Release note**:
```release-note
[openstack-cloud-controller-manager] Make volume mounts of openstack-cloud-controller-manager Helm Chart configurable
```

[1] Depending on the Kubernetes distribution (e.g. OCP, k3s etc.)
[2] For example, our Fedora CoreOS nodes simply don't have /usr/libexec/kubernetes/plugins/volume/exec and since the entire root filesystems of these nodes is read-only, the directory also cannot be created. Consequently, the entire mount fails.

### CODE REVIEW

File: charts/openstack-cloud-controller-manager/Chart.yaml
Comment (jichenjc) [inline charts/openstack-cloud-controller-manager/Chart.yaml:7]: why bump to 1.2? maybe need 1.1.2?

**Comment by jichenjc** [comment]
just one minor question, others seems fine

File: charts/openstack-cloud-controller-manager/Chart.yaml
Comment (jacksgt) [reply charts/openstack-cloud-controller-manager/Chart.yaml:7]: Hi, thanks for getting back to me so quickly.

I was following [Semantic versioning](https://semver.org/), which [should be used for Helm charts](https://helm.sh/docs/chart_best_practices/conventions/#version-numbers):

> Given a version number MAJOR.MINOR.PATCH, increment the:
>   MAJOR version when you make incompatible API changes,
>   MINOR version when you add functionality in a backwards compatible manner, and
>   PATCH version when you make backwards compatible bug fixes.

**Comment by jacksgt** [comment]
@jichenjc Hi, the CI test failed, but this does not seem to be cause by this PR, rather a test script seems to be missing:
```
 + ./tests/ci-keystone-e2e.sh
/usr/local/bin/runner.sh: line 114: ./tests/ci-keystone-e2e.sh: No such file or directory 
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/cloud-provider-openstack/1678/openstack-cloud-keystone-authentication-authorization-test/1456100003477983232#1:build-log.txt%3A17

**Comment by jichenjc** [comment]
right, the test scripts is supposed to be added later on 
recently a test-infra PR merged for e2e keystone, for now you we can ignore that failure
and overwrite it later on

I will submit a PR to make test pass first (maybe a fake stub first)
thanks for point that out

File: charts/openstack-cloud-controller-manager/Chart.yaml
Comment (lingxiankong) [reply charts/openstack-cloud-controller-manager/Chart.yaml:7]: Hi both, appreciate the discussion here. However, it has been decided (within the team) to only bump the chart version when we do release, so please still keep `1.1.1`.

@jichenjc remember? Please see the discussion in Slack. I just need to add some doc for contributors.

Otherwise, this PR looks good to me.

**Review by lingxiankong — CHANGES_REQUESTED** [review]

**Comment by jacksgt** [comment]
Hi @lingxiankong , as requested I have reset the version number to 1.1.1

**Comment by lingxiankong** [comment]
> Hi @lingxiankong , as requested I have reset the version number to 1.1.1

Thanks. Please ignore the chart lint job, we are fixing the failure.

**Comment by lingxiankong** [comment]
We have a fix the helm chart job: https://github.com/kubernetes/cloud-provider-openstack/pull/1691, please rebase the PR once that is merged.

**Comment by jacksgt** [comment]
> We have a fix the helm chart job: #1691, please rebase the PR once that is merged.

Done.

File: charts/openstack-cloud-controller-manager/templates/daemonset.yaml
Comment (strigazi) [inline charts/openstack-cloud-controller-manager/templates/daemonset.yaml:41]: Is this needed for the configurable volume mounts?

File: charts/openstack-cloud-controller-manager/values.yaml
Comment (strigazi) [inline charts/openstack-cloud-controller-manager/values.yaml:72]: What about leaving the extra mounts empty by default?

Like this, `/usr/libexec/kubernetes/kubelet-plugins/volume/exec` can not be removed.


Have you considered the approach below? To leave it empty and let users pick what they need?
https://github.com/kubernetes/ingress-nginx/blob/ab4fa4246797a33bc7b45b45097a91086fe9ad48/charts/ingress-nginx/values.yaml#L537-L545

Comment (jacksgt) [reply charts/openstack-cloud-controller-manager/values.yaml:72]: Leaving it empty would be a backwards-incompatible change.
You can override the entire array by specifying `extraVolumeMounts: []` (or adding the desired entries only).

File: charts/openstack-cloud-controller-manager/templates/daemonset.yaml
Comment (jacksgt) [reply charts/openstack-cloud-controller-manager/templates/daemonset.yaml:41]: This is not required for this, but the `address` flag has been deprecated. When using `--address`, occm prints the following:
```
Flag --address has been deprecated, see --bind-address instead.
```

Comment (strigazi) [reply charts/openstack-cloud-controller-manager/templates/daemonset.yaml:41]: I leave it to the approvers if they want to take this in on commit.

File: charts/openstack-cloud-controller-manager/values.yaml
Comment (strigazi) [reply charts/openstack-cloud-controller-manager/values.yaml:72]: It seems that it doesn't merge the default extraVolumeMounts and the user-defined ones. We can override the defaults, thanks!

Is this syntax used elsewhere? I still think it would be friendlier to use standard kubernetes syntax for volumes and volumeMounts.
```
  - name: flexvolume-dir
    hostPath: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
    mountPath: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
```

Comment (jacksgt) [reply charts/openstack-cloud-controller-manager/values.yaml:72]: What do you mean by "standard kubernetes syntax"?

Comment (strigazi) [reply charts/openstack-cloud-controller-manager/values.yaml:72]: In the examples below volumes and volumeMounts are defined as they would in kubernetes manifest. Not sure how else to explain "standard kubernetes syntax".

* https://github.com/kubernetes/cloud-provider-openstack/blob/a9da92cddac704db6d940abbc8f1b2dc20b673cc/charts/cinder-csi-plugin/values.yaml#L52-L62
* mentioned above https://github.com/kubernetes/ingress-nginx/blob/ab4fa4246797a33bc7b45b45097a91086fe9ad48/charts/ingress-nginx/values.yaml#L537-L545

Comment (jacksgt) [reply charts/openstack-cloud-controller-manager/values.yaml:72]: I see, thanks for the example. Well, I tried to adapt the data structure to this use-case because I don't like exposing all the implementation details (otherwise the user eventually needs to pass in the entire `spec` structure).

Comment (gman0) [reply charts/openstack-cloud-controller-manager/values.yaml:72]: Hi @jacksgt, thanks for your contribution! This really seems a bit over-complicated IMO. We'd need to keep around the comment documenting various volume types and options. We'd also lose input validation (imagine making a typo `subpath` instead of `subPath`).

I think it's best to stick with what's already in Kubernetes if we can. kubernetes/ingress-nginx seems like a good example to follow, having `extraVolumes` and `extraVolumeMounts` fields. Not only is this approach more flexible for users, we keep all the input validation. Maybe adding a comment referring users to the official Kubernetes docs on what's the syntax for volumes and volume mounts would help too.

Comment (jacksgt) [reply charts/openstack-cloud-controller-manager/values.yaml:83]: Alrighty, thanks for the feedback. I have updated the PR accordingly.
Would you mind taking a look? @gman0 @strigazi

Comment (gman0) [reply charts/openstack-cloud-controller-manager/values.yaml:83]: @jacksgt great, thank you!

**Comment by gman0** [comment]
Some changes were made. @lingxiankong @jichenjc can you PTAL?

**Comment by jichenjc** [comment]
/lgtm

leave to @lingxiankong for a look as he had some questions at first

**Comment by lingxiankong** [comment]
/approve

Thanks all the people involved.

**Pull request merged**
